content/mpi-and-threads-pt1.rst

```diff
@@ -3,31 +3,273 @@ Introducing MPI and threads
+
+
+Why move past pure-MPI programs?
+--------------------------------
+
+MPI is an excellent tool for parallel execution of programs. A key
+strength is that the programmer must explicitly move data to where it
+is needed. That can make code easier to understand, albeit more work
+to write. Since one writes code less often than the author *and
+maintainers* read it, that is often desirable.
+
+However, such approaches tend to perform poorly at scale. Computer
+hardware has not grown much faster in the last 20 years, but instead
+made many more potentially parallel execution units available. One can
+hardly buy a single-core CPU any more, and HPC nodes with multiple CPU
+sockets each containing scores of cores abound. MPI was designed in an
+era where it was much more common to find a node with only a single
+processor and a single core, and total counts were in the hundreds to
+thousands. Such jobs are often *too small* to be permitted to run on
+current high-end clusters.
+
+Code that assumes one MPI process to a core has trouble scaling to
+:math:`N` processes (eg. for N > 1000) for several reasons:
+
+* collective communication cost tends to scale at least like
+  :math:`\mathrm{log} N` - aggregating messages helps a lot but more
+  processes as starting and ending points for messages simply must
+  take more time
+
+* data must be replicated to each process, which takes time that grows
+  with the process count
+
+* replicated data forces cores to share the memory bandwith, thus
+  defeating the advantages of shared memory caches
+
+MPI + threading
+---------------
+  
+The MPI standard has been updated to accommodate the use of threads
+within processes. Using these capabilities is optional, and presents
+numerous advantages and disadvantages
+
+Advantages of MPI + threading
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* possiblity for better scaling of communication costs
+
+* either simpler and/or faster code that does not need to distribute
+  as much data, because all threads in the process can share it
+  already
+
+* higher performance from using memory caches better
+
+* reduced need to dedicate a rank solely to communication coordination
+  in code using a manager-worker paradigm
+
+* TODO
+
+Disadvantages of MPI + threading
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* implicitly shared data can be harder to reason about correctly
+  (eg. race conditions)
+
+* code now has to be correct MPI code *and* correct threaded code
+
+* possibility of lower performance from cache contention
+
+* more code complexity
+
+* might be merely shifting bottlenecks from one place to another
+  (eg. opening and closing OpenMP thread regions)
+
+* needs higher quality MPI implementations
+
+* TODO
+
+Threading library options
+-------------------------
+
+`OpenMP <https://www.openmp.org/>`_ is the open standard for HPC
+threading, and is widely used with many quality implementations. It is
+possible to use raw ``pthreads``, and you will find MPI examples using
```

Comment:
```suggestion
possible to use raw ```pthreads`` <https://en.wikipedia.org/wiki/POSIX_Threads>`_, and you will find MPI examples using
```